pytorch_lightning/callbacks/batch_size_finder.py

```diff
@@ -0,0 +1,319 @@
+
+        Hereafter, the batch size is further refined using a binary search
+        """
+        low = 1
+        high = None
+        count = 0
+        while True:
+            garbage_collection_cuda()
+            try:
+                # Try fit
+                self._try_loop_run(trainer)
+                count += 1
+                if count > self.max_trials:
+                    break
+                # Double in size
+                low = new_size
+                if high:
+                    if high - low <= 1:
+                        break
+                    midval = (high + low) // 2
+                    new_size, changed = self._adjust_batch_size(trainer, value=midval, desc="succeeded")
+                else:
+                    new_size, changed = self._adjust_batch_size(trainer, factor=2.0, desc="succeeded")
+
+                if changed:
+                    # Force the dataloaders to reset as the batch size has changed
+                    self._reset_dataloaders(trainer, pl_module)
+                else:
+                    break
+
+            except RuntimeError as exception:
+                # Only these errors should trigger an adjustment
+                if is_oom_error(exception):
+                    # If we fail in power mode, half the size and return
+                    garbage_collection_cuda()
+                    high = new_size
+                    midval = (high + low) // 2
+                    new_size, _ = self._adjust_batch_size(trainer, value=midval, desc="failed")
+                    if high - low <= 1:
+                        break
+                else:
+                    raise  # some other error not memory related
+
+        return new_size
+
+    def _try_loop_run(self, trainer):
+        if trainer.state.fn == TrainerFn.FITTING:
+            trainer.fit_loop.global_step = self._dumped_params["global_step"]
+            trainer.fit_loop.current_epoch = self._dumped_params["current_epoch"]
+            trainer.fit_loop.run()
+        elif trainer.state.fn == TrainerFn.VALIDATING:
+            trainer.validate_loop.run()
+        elif trainer.state.fn == TrainerFn.TESTING:
+            trainer.test_loop.run()
+        elif trainer.state.fn == TrainerFn.PREDICTING:
+            trainer.predict_loop.run()
+
+    @staticmethod
+    def _reset_dataloaders(trainer, pl_module):
+        if trainer.state.fn == TrainerFn.FITTING:
+            trainer.reset_train_dataloader(pl_module)
+            trainer.reset_val_dataloader(pl_module)
+        elif trainer.state.fn == TrainerFn.VALIDATING:
+            trainer.reset_val_dataloader(pl_module)
+        elif trainer.state.fn == TrainerFn.TESTING:
+            trainer.reset_test_dataloader(pl_module)
+        elif trainer.state.fn == TrainerFn.PREDICTING:
+            trainer.reset_predict_dataloader(pl_module)
+
+    def _dump_params(self, trainer):
+        self._dumped_params = {
+            "current_epoch": trainer.current_epoch,
```

Comment:
Might be better to deecopy the loops

Reply:
using loop state dict now. I guess that should work??